### ISSUE TITLE
Hello World => Segmentation fault

### ISSUE THREAD

**Issue by mcandre** [NuxiNL/cloudlibc#18]
Hello, I am trying to write my first CloudABI program, following along with the tutorials. I have a Clang 6.0-based CloudABI toolchain installed, along with cloudabi-run. Here is my source:

```c
#include <stdio.h>

int main(void) {
    dprintf(1, "Hello World!\n");
}
```

Trace:

```console
$ x86_64-unknown-cloudabi-cc -o hello hello.c
$ cloudabi-run hello
Segmentation fault
$ ./hello
Segmentation fault
```

Finally, here is my setup script for Docker, in case that matters:

https://github.com/mcandre/cartel/blob/cloudabi/setup-cloudabi.ubuntu.sh

**Comment by EdSchouten** [comment]
Yes. I agree the site needs an overhaul. Especially to deal with https://github.com/NuxiNL/cloudabi/issues/10. I'll try to work on this soon, also incorporating the suggestions you made here.

To respond to this question:

> A larger question is why are guarded? I thought CloudABI was meant to protect sensitive components that can break a system. Do stdout/stderr somehow contribute to an increased attack surface?

They are not necessarily an attack surface, but there's also some cohesion/usability/orthogonality concerns to it.

The main argument for omitting stdin/stdout/stderr is that they overlap with how `program_main()` / `program_exec()` / `program_spawn()` work. Those functions try to pack the entire configuration that needs to be passed from one process to another in a single argdata tree. If we also supported stdin/stdout/stderr in the traditional sense, the functions could be extended to pass those in separately.

```c
int program_exec(int program, const argdata_t *ad, int stdin, int stdout, int stderr);
```

But this is a slipperly slope: what would happen if someone comes along, saying `program_exec()` should also have a dedicated parameter to pass in a hypothetical 'PRNG file descriptor' to override the behaviour of `cloudabi_sys_random_get()`? What about passing in 'userspace clock file descriptors' to override the process's observation of time? Soon we'd have a version of `program_exec()` taking 10 file descriptor arguments.

```c
int program_exec(int program, const argdata_t *ad, int stdin, int stdout, int stderr, int clock_realtime, int clock_monotonic, int prng, [...]);
```

What I'd rather like to see is that if such features are ever needed, they are implemented without growing the API. They can all be passed in using argdata. If needed, we can work towards standards/conventions on what the schema of the argdata may look like.

There's also the issue with file descriptor numbering. On UNIX systems there is already the issue that [programs misbehave when launching them without file descriptors 0, 1 and 2 in place](https://www.oreilly.com/library/view/secure-programming-cookbook/0596003943/ch01s05.html). The first files opened by the child process will end up having one of those numbers, meaning an `printf()` will end up writing garbage to a random file. I guess that this would become an even bigger problem for us, as CloudABI is more file descriptor centric (as opposed to path/namespace centric) than traditional UNIX.

Finally, omitting designated stdin/stdout/stderr file descriptors was also done to remove the mindset that only a single input/output source/sink exists. There's absolutely nothing against having multiple log streams for different subsystems in your application.
